cardano-api/src/Cardano/Api/LedgerState.hs

```diff
@@ -0,0 +1,849 @@
+pattern LedgerStateMary st <- LedgerState  (Consensus.LedgerStateMary st)
+
+{-# COMPLETE LedgerStateByron
+           , LedgerStateShelley
+           , LedgerStateAllegra
+           , LedgerStateMary #-}
+
+data FoldBlocksError
+  = FoldBlocksInitialLedgerStateError InitialLedgerStateError
+  | FoldBlocksApplyBlockError Text
+
+renderFoldBlocksError :: FoldBlocksError -> Text
+renderFoldBlocksError fbe = case fbe of
+  FoldBlocksInitialLedgerStateError err -> renderInitialLedgerStateError err
+  FoldBlocksApplyBlockError err -> "Failed when applying a block: " <> err
+
+-- | Monadic fold over all blocks and ledger states. Stopping @k@ blocks before
+-- the node's tip where @k@ is the security parameter.
+foldBlocks
+  :: forall a.
+  FilePath
+  -- ^ Path to the cardano-node config file (e.g. <path to cardano-node project>/configuration/cardano/mainnet-config.json)
+  -> FilePath
+  -- ^ Path to local cardano-node socket. This is the path specified by the @--socket-path@ command line option when running the node.
+  -> NetworkId
+  -- ^ The network ID.
+  -> Bool
+  -- ^ True to enable validation. Under the hood this will use @applyBlock@
+  -- instead of @reapplyBlock@ from the @ApplyBlock@ type class.
+  -> a
+  -- ^ The initial accumulator state.
+  -> (Env -> LedgerState -> BlockInMode CardanoMode -> a -> IO a)
+  -- ^ Accumulator function Takes:
+  --  * Environment (this is a constant over the whole fold)
+  --  * The current Ledger state (with the current block applied)
+  --  * The current Block
+  --  * The previous state
+  --
+  -- And this should return the new state.
+  --
+  -- Note: This function can safely assume no rollback will occur even though
+  -- internally this is implemented with a client protocol that may require
+  -- rollback. This is achieved by only calling the accumulator on states/blocks
+  -- that are older than the security parameter, k. This has the side effect of
+  -- truncating the last k blocks before the node's tip.
+  -> ExceptT FoldBlocksError IO a
+  -- ^ The final state
+foldBlocks nodeConfigFilePath socketPath networkId enableValidation state0 accumulate = do
+  -- NOTE this was originally implemented with a non-pipelined client then
+  -- changed to a pipelined client for a modest speedup:
+  --  * Non-pipelined: 1h  0m  19s
+  --  * Pipelined:        46m  23s
+
+  (env, ledgerState) <- withExceptT FoldBlocksInitialLedgerStateError
+                            (initialLedgerState nodeConfigFilePath)
+
+  -- Place to store the accumulated state
+  -- This is a bit ugly, but easy.
+  errorIORef <- lift $ newIORef Nothing
+  stateIORef <- lift $ newIORef state0
+
+  -- Connect to the node.
+  lift $ connectToLocalNode
+    connectInfo
+    (protocols stateIORef errorIORef env ledgerState)
+
+  (lift $ readIORef errorIORef) >>= \case
+    Just err -> throwE (FoldBlocksApplyBlockError err)
+    Nothing -> lift $ readIORef stateIORef
+  where
+  connectInfo :: LocalNodeConnectInfo CardanoMode
```

Comment:
Lets put an indent here to align with the coding style of the repo. This will make the `where` clause stand out more.